Add support for nomad-native services

t0xicCode opened the pull request:
I decided to add support for connecting gocast to a nomad cluster to pull services from it.

Still need to polish and add documentation.

t0xicCode committed:
Add support for getting services out of nomad

t0xicCode committed:
Add support for getting services out of nomad

t0xicCode commented:
@mayuresh82 is there a preference for specifying which node gocast should care about? For the consul side, it’s done with an environment variable, whereas I’ve initially done it with a configuration file parameter. I’m happy to switch it to env variable for consistency.

t0xicCode committed:
Update messages

t0xicCode committed:
do not send communities if none are defined

t0xicCode committed:
Use local IP if available event for 3-part nat

t0xicCode committed:
nomad: make addr and node required for monitor

t0xicCode committed:
Add documentation for the nomad integration

t0xicCode commented:
should be ready for review/merging

t0xicCode committed:
Add logging for nat rules

mayuresh82 commented:
Nomad is not really a service discovery solution so wondering what is the motivation for this ?

t0xicCode commented:
Nomad Services can be exposed directly/natively within nomad without using a consul instance, it now has native service discovery. This enables that use case.

See https://developer.hashicorp.com/nomad/docs/networking/service-discovery

Vaelatern commented:
I use nomad for service discovery of services running within nomad, and am prepared to test this PR.

t0xicCode commented:
@Vaelatern feel free to test it using my fork, I’ve been running it for a year without issues.

mayuresh82 commented:
@t0xicCode  thanks for the contribution. Happy to merge this in once the below is addressed. TY !
- can you please provide a test plan here , showing logs talking to nomad, announcing withdrawing VIPs based on nomad healthcheck etc. 
- Add unit test for nomad.go
- cleanup the unwanted files ( i see a binary file in there )
-

t0xicCode committed:
remove binary file

t0xicCode commented:
@mayuresh82 

I've added a unit test for `nomad.go`. I've also removed the `gocast` binary. The current repository includes a `gocast` binary (built 4 years ago), so I assumed that it was expected to have a binary for amd64 present. Separately, I'd be happy to contribute a goreleaser-based github action to build the code on release/commit if you're interested.

As for a test plan, I've included some logs showing what happens when a service is removed by nomad itself (which happens if a healthcheck fails or the allocation associated with the service is relocated to a new node):

Using the following tags on the service
```
enable_gocast
gocast_vip=172.31.255.1/32
gocast_nat=tcp:443
```

Node 1
```
I0317 18:03:08.183863       1 nomad.go:100] Querying nomad service at http://nomad.server///v1/services?namespace=%2A
I0317 18:03:08.187296       1 nomad.go:154] Querying nomad service acme at http://nomad.server///v1/service/acme?namespace=internal&filter=NodeID+%3D%3D+%223d1681d9-1e62-5824-485c-47a15e7216ea%22
I0317 18:03:08.189621       1 nomad.go:180] found service acme (ns: internal) for this node
I0317 18:03:08.189669       1 monitor.go:181] App acme@internal already exists
I0317 18:03:13.118758       1 monitor.go:283] All Monitors for app: acme@internal succeeded
I0317 18:03:23.117011       1 monitor.go:283] All Monitors for app: acme@internal succeeded
I0317 18:03:33.118769       1 monitor.go:283] All Monitors for app: acme@internal succeeded
I0317 18:03:38.194098       1 nomad.go:100] Querying nomad service at http://nomad.server///v1/services?namespace=%2A
I0317 18:03:38.196741       1 nomad.go:154] Querying nomad service acme at http://nomad.server///v1/service/acme?namespace=internal&filter=NodeID+%3D%3D+%223d1681d9-1e62-5824-485c-47a15e7216ea%22
I0317 18:03:38.198729       1 nomad.go:66] Removing app: acme@internal as it was not found in nomad
I0317 18:03:38.198849       1 monitor.go:356] Exit run-loop for app: acme@internal
time="2026-03-17T18:03:38-04:00" level=debug msg="Executing: iptables -t nat -D PREROUTING -p tcp -d 172.31.255.1 --dport 443 -j DNAT --to-destination 100.64.0.2:28348"
I0317 18:03:38.205438       1 system.go:98] updating nat rules: iptables -t nat -D PREROUTING -p tcp -d 172.31.255.1 --dport 443 -j DNAT --to-destination 100.64.0.2:28348
```

Node 2:
```
I0317 18:03:26.105689       1 nomad.go:100] Querying nomad service at http://nomad.server///v1/services?namespace=%2A
I0317 18:03:26.109020       1 nomad.go:154] Querying nomad service acme at http://nomad.server///v1/service/acme?namespace=internal&filter=NodeID+%3D%3D+%2245f59709-1943-1577-e172-89fbab1f40d0%22
I0317 18:03:56.131567       1 nomad.go:100] Querying nomad service at http://nomad.server///v1/services?namespace=%2A
I0317 18:03:56.134558       1 nomad.go:154] Querying nomad service acme at http://nomad.server///v1/service/acme?namespace=internal&filter=NodeID+%3D%3D+%2245f59709-1943-1577-e172-89fbab1f40d0%22
I0317 18:03:56.136961       1 nomad.go:180] found service acme (ns: internal) for this node
I0317 18:03:56.137181       1 monitor.go:342] Starting run-loop for app acme@internal
I0317 18:03:56.137232       1 monitor.go:283] All Monitors for app: acme@internal succeeded
time="2026-03-17T18:03:56-04:00" level=debug msg="not announced, announcing acme@internal"
I0317 18:03:56.137193       1 monitor.go:203] Registered a new app: Name: acme@internal, Vip: 172.31.255.1/32, VipConf: {[]}, Monitors: [], Nats: [tcp:443], Source: nomad
time="2026-03-17T18:03:56-04:00" level=debug msg="Executing: iptables -t nat -A PREROUTING -p tcp -d 172.31.255.1 --dport 443 -j DNAT --to-destination 100.64.0.3:25892"
I0317 18:03:56.147724       1 system.go:98] updating nat rules: iptables -t nat -A PREROUTING -p tcp -d 172.31.255.1 --dport 443 -j DNAT --to-destination 100.64.0.3:25892
time="2026-03-17T18:03:56-04:00" level=info msg="Add a peer configuration" Key=192.168.50.1 Topic=Peer
time="2026-03-17T18:03:57-04:00" level=info msg="skipped asn negotiation" Asn=64512 Key=192.168.50.1 PeerType=external State=BGP_FSM_OPENSENT Topic=Peer
time="2026-03-17T18:03:57-04:00" level=info msg="Peer Up" Key=192.168.50.1 State=BGP_FSM_OPENCONFIRM Topic=Peer
I0317 18:04:06.160030       1 monitor.go:283] All Monitors for app: acme@internal succeeded
I0317 18:04:16.165167       1 monitor.go:283] All Monitors for app: acme@internal succeeded
I0317 18:04:26.137628       1 nomad.go:100] Querying nomad service at http://nomad.server///v1/services?namespace=%2A
I0317 18:04:26.143628       1 nomad.go:154] Querying nomad service acme at http://nomad.server///v1/service/acme?namespace=internal&filter=NodeID+%3D%3D+%2245f59709-1943-1577-e172-89fbab1f40d0%22
I0317 18:04:26.146519       1 nomad.go:180] found service acme (ns: internal) for this node
I0317 18:04:26.146576       1 monitor.go:181] App acme@internal already exists
```

t0xicCode committed:
Add tests for nomad

t0xicCode committed:
replace app when the endpoint nomad changes

mayuresh82 commented:
Thanks. This is awesome. You are welcome to contribute a Goreleaser spec for automated release,  that will be  great.

mayuresh82 commented on controller/app.go:
```diff
@@ -56,6 +56,8 @@ type App struct {
 	Monitors  Monitors
 	Nats      []string
+	Addr      net.IP
+	Port      int
```
please clarify the significance of Addr and Port. What exactly do they mean in the context of an App ?

Vaelatern replied:
These are the address and port where the program is bound. On workers with multiple IP addresses, the service is only going to end up bound to one Addr:Port pair.

t0xicCode replied:
Added some comments  to explain the use of these fields (and wrapped them in a struct).

mayuresh82 commented on controller/app.go:
```diff
@@ -68,7 +70,14 @@ func (a *App) Equal(other *App) bool {
+}
+
+// EndpointEqual determines if the Addr and Port of an App matches this one's.
+// This is only useful for the nomad apps as the consul ones don't use those fields.
```
Since the concept of `Endpoint` is nomad specific,  a suggestion is to add a couple of enums to `App` to indicate the source of the app
```
type AppSource string
const (
  nomadApp AppSource = "nomadApp"
  consulApp AppSource = "consulApp"
)
type App struct {
    source AppSource
}
```
then you can check for `EndpointChanged` only if its a nomad app below:
```
   if existing.app.source == nomadAPP && !existing.app.EndpointEqual(app) {
			// the endpoint changed!
```

Vaelatern replied:
fwiw I even went further in my own attempt, which I never got working, doing Nomad more.

Vaelatern commented:
I'm trying to work out running Gocast in Nomad, with NOMAD_TOKEN taken from the job's identity{} and the NOMAD_ADDR as a UNIX socket. This has proven slightly tricky. To support this my fork of this repo has a github action to release a container with gocast built inside.

t0xicCode commented:
> I'm trying to work out running Gocast in Nomad, with NOMAD_TOKEN taken from the job's identity{} and the NOMAD_ADDR as a UNIX socket. This has proven slightly tricky. To support this my fork of this repo has a github action to release a container with gocast built inside.

Once this is accepted, I'd love to see your thoughts/changes to support unix sockets!

Vaelatern commented:
What about considering getting the node_id from /v1/agent/self from the API?

t0xicCode commented:
> What about considering getting the node_id from /v1/agent/self from the API?

Oh that’s a good idea, I’ll give it a shot.

t0xicCode committed:
sort imports

t0xicCode committed:
update go version requirements

t0xicCode committed:
nomad: add support for unix socket

t0xicCode committed:
do not send communities if none are defined

t0xicCode commented:
@Vaelatern I added support for Unix sockets and reading the node ID from the API

@mayuresh82 happy to help doing code-review/maintaining if you're still looking for assistance.

Vaelatern commented:
The nomad socket stuff works.

But the NAT stuff does not. I'm thinking the addition of `App.Endpoint` was a mistake, and the data should have been shoehorned into the original format.

I'm also concerned that it's normal to put multiple nomad services on a single task, but it's also normal to expect a VIP to answer on many ports (not having a different IP address for e.g. http and https). I'm mulling over what kind of abstraction would help advertise "groups" of services under the same VIP, or even just NATting all traffic to a VIP to a specific task....

t0xicCode commented:
@Vaelatern what error(s) do you get? It's working in my deployment.

Regarding the port issue, I went for the initial contract of one vip per service. I think that anything else would require a bigger refactor and would likely also require some consideration on the consul side of things.

Vaelatern commented:
My error is the actual iptables rules are forwarding traffic to the dport not to the actual endpoint's listening port. 

I'll draft a patch to this patch set

Vaelatern commented:
Working on testing https://github.com/Vaelatern/gocast/commit/2c7bca751d796f3c65dcba1f41841955c63f155f to fix the nat issue

It doesn't yet handle the replacement case correctly either...

t0xicCode commented:
Removing the Endpoint should be fine, but we’d need to make sure that we
replace the app if the Nat rules don’t match.

If a task is rescheduled on the same host, the gocast instance needs to
update the bat rules to account for the new dport. I’ll work on something

> <https://github.com/mayuresh82/gocast/pull/29#issuecomment-4911980082>
>

Vaelatern commented:
I think the right place to merge NAT rules is on the gocast_vip - but am tempted to add a "merge-with-$name" idea, so we can say "Here is another app, but it's actually the same as this other app, please merge them so the announce is the same"

t0xicCode committed:
Migrate NAT rules to custom chain

Moves the NAT rules to a custom chain (GOCAST).
This removes pollution in the built-in chains and allows for reuse of
rules, which makes enabling hairpin easy.

t0xicCode committed:
Switch up how nomad services figure out which IP to use as destination

t0xicCode commented:
@Vaelatern I took another look at both the merging of rules and the detection of the service address for nomad.

On the merging front, if an app is equal but has different NAT rules, then the existing app gets removed and the new one gets added in its place.

On the service detection, I added an undocumented format for `gocast_nat`: `protocol:vipPort:destinationIP:destinationPort`. If the `destinationIP` field is missing, it gets populated by the IP address of the nomad service.

Vaelatern commented on controller/monitor.go:
```diff
@@ -170,53 +172,60 @@ func (m *MonitorMgr) consulMon() {
-	for _, appMon := range m.monitors {
+	// check if already running
+	for appMon := range m.Monitors() {
+		// TODO properly handle a situation where multiple apps try to claim the same ip/port pair
```
they can't claim the same ip:port, it's prevented by the OS.

Vaelatern commented on controller/monitor.go:
```diff
@@ -237,33 +246,29 @@ func (m *MonitorMgr) Remove(appName string) {
+			case 3:
+				// protocol:vipPort:servicePort
+				if err := natRule(iptables.Delete, a.app.Vip.Net.IP, m.ctrl.localIP, parts[0], parts[1], parts[2]); err != nil {
+					glog.Errorf("Failed to remove app %s: %v", a.app.FullName(), err)
```
Why would we ever NAT to an IP that's not the host IP where the services are listening? gocast already only advertises for its own IP - this addition just to not throw out the IP information seems maybe not necessary?

t0xicCode replied:
With multiple network interfaces, you can have a situation where the nomad service listens on a different network interface.

e.g. I run my services such that they only listen on a single ip on one network interface (e.g. `eth1`, `10.0.0.10`), but the bgp session (and thus the local IP) is on another network interface (e.g. `eth0`, `192.168.1.10`).

By tracking the IP to which the nomad service is bound, gocast can create the nat rule to redirect traffic without requiring the service to listen on `0.0.0.0`.

Vaelatern commented on controller/nomad.go:
```diff
@@ -17,9 +17,9 @@ import (
-	nomadSecretsDirEnv = "NOMAD_SECRETS_DIR"
-	nomadTokenEnv      = "NOMAD_TOKEN"
+	defaultNomadAddr = "http://127.0.0.1:4646"
+	nomadUnixAddrEnv = "NOMAD_UNIX_ADDR"
```
edit: Resolved, is good.
Nomad [documentation](https://developer.hashicorp.com/nomad/api-docs/task-api) states:

```
The Unix Domain Socket is located at ${NOMAD_SECRETS_DIR}/api.sock.
```

`NOMAD_UNIX_ADDR` is not set by default. This is a needless change, `NOMAD_SECRETS_DIR` is sufficient.

t0xicCode replied:
`NOMAD_UNIX_ADDR` should be set. [Documentation](https://developer.hashicorp.com/nomad/docs/reference/runtime-environment-settings) states:

> | Variable | Description |
> | --- | --- |
> | `NOMAD_UNIX_ADDR` | Use this value as your `NOMAD_ADDR` to use nomad CLI with the [task API](https://developer.hashicorp.com/nomad/api-docs/task-api)'s unix socket. The value is equivalent to "unix://${NOMAD_SECRETS_DIR}/api.sock" |

This reads to me as "`NOMAD_UNIX_ADDR` should be preferred".

Vaelatern replied:
Indeed, you have read the documentation better than I!

Vaelatern commented on controller/nomad.go:
```diff
@@ -78,8 +78,8 @@ func (c *nomadClient) Do(req *http.Request) (resp *http.Response, err error) {
-		if dir := os.Getenv(nomadSecretsDirEnv); dir != "" {
-			addr = fmt.Sprintf("unix://%s/api.sock", dir)
+		if unixAddr := os.Getenv(nomadUnixAddrEnv); unixAddr != "" {
+			addr = unixAddr
```
edit: Resolved, is good.
Should revert per previous comment on R21.

Vaelatern commented on controller/app.go:
```diff
@@ -155,10 +130,24 @@ func (a *App) Equal(other *App) bool {
 		a.Source == other.Source
```
on nomad services that are the same app this is not the case because each service{} gets its own name.

t0xicCode replied:
What do you mean here?

Vaelatern replied:
https://developer.hashicorp.com/nomad/docs/job-specification/service#name

I meant the statement I remarked on, not the exact line. There can not be two service{} blocks where a.Name == other.Name unless they are the identical service block. Thus it's useless to check when merging -- when merging nomad jobs at least.

t0xicCode replied:
The way the services are added, it tries to add the service to gocast every time. The gocast controller checks if there’s already a service which matches (based on the `Equal` function) and does a replacement if said service was updated.

So we want to compare the name, source, and vip info to either dismiss the new app, replace and existing one, or add a new one.

Vaelatern replied:
How does this fit with merging services into a single shared perspective with multiple distinct names?

"if an app is equal but has different NAT rules"

They can't be equal if their names are distinct.

t0xicCode replied:
Sorry, I'll try reformulating.

`NomadMonitor.Monitor` gets all a list of all the services running on the current nomad instance (from `NomadMonitor.queryServices`). Next, it calls `MonitorMgr.Add` for each app (service) that was seen. `MonitorMgr.Add` checks to see if there is an app which matches (name, source, vip). If so, it checks to see if the nat rules changed and replaces the app if that's the case. If the app doesn't exist, it adds it.

So on every monitor loop, apps will be generated that may match existing apps.

Vaelatern replied:
so what's the useful solution for checking equality? if source is nomad, don't check the name when checking equality? Especially with merging.

t0xicCode replied:
I'm not sure I understand what you mean?

`Equal` only compares a subset of attributes which prevents duplicates and makes app updates work.

mayuresh82 commented:
is this ready for review @t0xicCode  ?

t0xicCode commented:
Ready for review!
